src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,254 @@
+    type Output = &'static mut [u8];
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // TODO: Detect when the print is done, retrieve `buffer`, and return
+        // it.
+    }
+}
+```
+
+Note: All error handling is omitted to keep things understandable.
+
+Barbara begins to implement `print_buffer`:
+
+```rust
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
+    kernel_set_print_callback(callback);
+    kernel_start_print(buffer);
+    PrintFuture {}
+}
+
+// New! The callback the kernel calls.
+extern fn callback() {
+    // TODO: Wake up the currently-waiting PrintFuture.
+}
+```
+
+So far so good. Barbara then works on `poll`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            Poll::Ready(kernel_get_buffer_back())
+        }
+        Poll::Pending
+    }
+```
+
+Of course, there's something missing here. How does the callback wake the
+`PrintFuture`? She needs to store the
+[`Waker`](https://doc.rust-lang.org/stable/core/task/struct.Waker.html)
+somewhere! Barbara puts the `Waker` in a global variable so the callback can
+find it (this is fine because the app is single threaded and callbacks do NOT
+interrupt execution the way Unix signals do):
+
+```rust
+static mut PRINT_WAKER: Option<Waker> = None;
+
+extern fn callback() {
+    if let Some(waker) = unsafe { PRINT_WAKER.as_ref() } {
+        waker.wake_by_ref();
+    }
+}
+```
+
+She then modifies `poll` to set `PRINT_WAKER`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            return Poll::Ready(kernel_get_buffer_back());
+        }
+        unsafe { PRINT_WAKER = Some(cx.waker()); }
+        Poll::Pending
+    }
+```
+
+`PRINT_WAKER` is stored in `.bss`, which occupies space in RAM but not flash. It
+is two words in size. It points to a
+[`RawWakerVTable`](https://doc.rust-lang.org/stable/core/task/struct.RawWakerVTable.html)
+that is provided by the executor. `RawWakerVTable` does not appear to be
+designed for a no-`alloc` environment. In a no-`alloc` environment, `drop` and
+`clone` are generally no-ops, and `wake`/`wake_by_ref` seem like duplicates.
+Looking at `RawWakerVTable` made Barbara think about the size impact of using
+`Future`.
+
+Barbara decided the overhead would be worth it if it made it easy to work with
+future combinators from external crates *and* could function as a lightweight
+form of multithreading. To support this, Barbara decides to implement an
+executor designed to function like a background thread. Because `alloc` is not
+available, this executor contains a single future. The executor has a `spawn`
+function that accepts a future and starts running that future (overwriting the
+existing future in the executor if one is already present). Once started, the
+executor runs entirely in kernel callbacks.
```

Comment:
> In general `Future`s assume they are only called by one executor in a serial fashion, and not from a variety of places. Doing the latter would void all `Send/Sync` guarantees.

Everything happens in a single thread, so I don't see how `Send`/`Sync` are relevant.

> And not only this, it would open up the door to a variety of reentrancy problems that the Future design avoids.

The executor described here will not call `poll` reentrantly. I don't see anywhere else reentrancy problems can occur.

Reply:
One example is that a `Future` which is `!Send` can expect all state which is stored in thread-local storage to be consistently accessible after polls. If one calls a future from arbitrary threads that is no longer true. In general an interrupt context is similar to a separate thread. I'm not familiar enough with tock to say whether this applies or doesn't apply there.

Reply:
> In general an interrupt context is similar to a separate thread. I'm not familiar enough with tock to say whether this applies or doesn't apply there.

Oh, that doesn't apply in Tock. Invoking a callback behaves almost identically to a virtual function call whose address is determined by the kernel.